client/api/src/leaves.rs

```diff
@@ -391,6 +374,20 @@ mod tests {
 		assert_eq!(set, set2);
 	}
 
+	#[test]
+	fn undo_import() {
+		let mut set = LeafSet::new();
+		set.import(10_1u32, 10u32, 0u32);
+		set.import(11_1, 11, 10_1);
+		set.import(11_2, 11, 10_1);
+
+		let displaced = set.import(12_1, 12, 11_2).unwrap();
+		assert!(set.contains(12, 12_1));
+
+		set.undo().undo_import(displaced);
+		assert!(!set.contains(12, 12_1));
```

Comment:
Maybe exhaustively check the number of elements + whole contents of the set?

Reply:
Yeah, I'll improve the finalization undo as well.

Reply:
Thanks for the hint. This made me spot a couple of bugs :-)